File: web/lib/EARNINGS_STANDARD.ts
/**
 * EARNINGS STANDARD - Single Source of Truth
 * 
 * This file defines the canonical earning structure for the entire application.
 * ALL earning calculations across frontend, backend, and database MUST reference
 * these values to ensure consistency.
 * 
 * Last Updated: 2025-01-11
 * Version: 1.0.0
 */

// ============================================
// TYPE DEFINITIONS
// ============================================

export type SpotterTier = 'elite' | 'verified' | 'learning' | 'restricted';
export type EarningType = 'trend_submission' | 'trend_validation' | 'approval_bonus' | 'scroll_session' | 'challenge' | 'referral';
export type EarningStatus = 'pending' | 'approved' | 'paid' | 'cancelled';

export interface EarningCalculation {
  baseAmount: number;
  bonusAmount: number;
  tierMultiplier: number;
  streakMultiplier: number;
  finalAmount: number;
  appliedBonuses: string[];
  breakdown: EarningBreakdown;
}

export interface EarningBreakdown {
  base: number;
  qualityBonuses: Record<string, number>;
  performanceBonuses: Record<string, number>;
  multipliers: {
    tier: number;
    streak: number;
  };
  total: number;
  capped: boolean;
}

export interface TrendSubmissionData {
  // Required fields
  trendName: string;
  description: string;
  
  // Quality bonus fields
  screenshot_url?: string;
  ageRanges?: any[];
  subcultures?: any[];
  otherPlatforms?: any[];
  creator_handle?: string;
  hashtags?: string[];
  post_caption?: string;
  
  // Performance bonus fields
  views_count?: number;
  likes_count?: number;
  comments_count?: number;
  wave_score?: number;
  category?: string;
  tickers?: string[];
  isFinanceTrend?: boolean;
  
  // Platform data
  platform?: string;
}

// ============================================
// EARNING CONFIGURATION CONSTANTS
// ============================================

export const EARNINGS_STANDARD = {
  // Version for tracking changes
  VERSION: '1.0.0',
  
  // Base earning rates (in USD)
  BASE_RATES: {
    TREND_SUBMISSION: 1.00,      // Base payment for submitting a trend
    VALIDATION_VOTE: 0.10,        // Payment per validation vote
    APPROVAL_BONUS: 0.50,         // Bonus when trend gets approved (paid to spotter)
    SCROLL_SESSION: 0.00,         // Scroll sessions don't pay directly (only maintain streaks)
  },
  
  // Quality bonuses - applied when submitting trend
  QUALITY_BONUSES: {
    SCREENSHOT: 0.15,             // Includes screenshot/media
    COMPLETE_INFO: 0.10,          // Has both title and detailed description
    DEMOGRAPHICS: 0.10,           // Includes age range data
    SUBCULTURES: 0.10,            // Tags relevant subcultures
    MULTI_PLATFORM: 0.10,         // Seen on multiple platforms
    CREATOR_INFO: 0.05,           // Includes creator handle
    RICH_HASHTAGS: 0.05,          // Has 3+ relevant hashtags
    CAPTION_PROVIDED: 0.05,       // Includes original post caption
  },
  
  // Performance bonuses - based on trend metrics
  PERFORMANCE_BONUSES: {
    VIRAL_CONTENT: 0.50,          // 1M+ views
    HIGH_VIEWS: 0.25,             // 100k-999k views (not cumulative with VIRAL)
    HIGH_ENGAGEMENT: 0.20,        // >10% engagement rate (likes/views)
    HIGH_WAVE_SCORE: 0.20,        // Wave score > 70
    FINANCE_TREND: 0.10,          // Finance/crypto/stocks related
    TRENDING_CATEGORY: 0.10,      // In a currently trending category
  },
  
  // Spotter tier multipliers (applied to final amount)
  TIER_MULTIPLIERS: {
    elite: 1.5,                   // 50% bonus for elite spotters
    verified: 1.0,                // Standard rate for verified
    learning: 0.7,                // 30% reduction for learning
    restricted: 0.3,              // 70% reduction for restricted
  },
  
  // Streak multipliers (cumulative with tier multiplier)
  STREAK_MULTIPLIERS: {
    0: 1.0,                       // No streak
    1: 1.0,                       // First submission
    2: 1.2,                       // 20% bonus
    3: 1.5,                       // 50% bonus
    5: 2.0,                       // 2x multiplier
    10: 2.5,                      // 2.5x multiplier
    15: 3.0,                      // 3x multiplier (max)
  },
  
  // System limits and caps
  LIMITS: {
    MAX_SINGLE_SUBMISSION: 3.00,  // Maximum earning per submission
    MAX_DAILY_EARNINGS: 50.00,    // Maximum daily earnings
    MIN_CASHOUT_AMOUNT: 5.00,     // Minimum balance for cashout
    STREAK_WINDOW_MINUTES: 5,     // Minutes to maintain streak
  },
  
  // Validation requirements
  VALIDATION: {
    VOTES_TO_APPROVE: 2,          // Votes needed to approve trend
    VOTES_TO_REJECT: 2,           // Votes needed to reject trend
    MAX_VOTE_EARNINGS_DAILY: 10.00, // Max daily earnings from voting
  },
  
  // Processing settings
  PROCESSING: {
    CASHOUT_HOURS: '24-48',       // Time to process cashouts
    PAYMENT_METHODS: ['venmo', 'paypal', 'bank_transfer'],
  },
} as const;

// ============================================
// CALCULATION FUNCTIONS
// ============================================

/**
 * Calculate earnings for a trend submission
 */
export function calculateTrendSubmissionEarnings(
  data: TrendSubmissionData,
  spotterTier: SpotterTier = 'learning',
  streakCount: number = 0
): EarningCalculation {
  const breakdown: EarningBreakdown = {
    base: EARNINGS_STANDARD.BASE_RATES.TREND_SUBMISSION,
    qualityBonuses: {},
    performanceBonuses: {},
    multipliers: {
      tier: EARNINGS_STANDARD.TIER_MULTIPLIERS[spotterTier],
      streak: getStreakMultiplier(streakCount),
    },
    total: 0,
    capped: false,
  };
  
  const appliedBonuses: string[] = [];
  
  // Apply quality bonuses
  if (data.screenshot_url) {
    breakdown.qualityBonuses.screenshot = EARNINGS_STANDARD.QUALITY_BONUSES.SCREENSHOT;
    appliedBonuses.push('📸 Screenshot');
  }
  
  if (data.trendName && data.description && data.description.length > 20) {
    breakdown.qualityBonuses.completeInfo = EARNINGS_STANDARD.QUALITY_BONUSES.COMPLETE_INFO;
    appliedBonuses.push('📝 Complete Info');
  }
  
  if (data.ageRanges && data.ageRanges.length > 0) {
    breakdown.qualityBonuses.demographics = EARNINGS_STANDARD.QUALITY_BONUSES.DEMOGRAPHICS;
    appliedBonuses.push('👥 Demographics');
  }
  
  if (data.subcultures && data.subcultures.length > 0) {
    breakdown.qualityBonuses.subcultures = EARNINGS_STANDARD.QUALITY_BONUSES.SUBCULTURES;
    appliedBonuses.push('🎭 Subcultures');
  }
  
  if (data.otherPlatforms && data.otherPlatforms.length > 0) {
    breakdown.qualityBonuses.multiPlatform = EARNINGS_STANDARD.QUALITY_BONUSES.MULTI_PLATFORM;
    appliedBonuses.push('🌐 Multi-Platform');
  }
  
  if (data.creator_handle) {
    breakdown.qualityBonuses.creatorInfo = EARNINGS_STANDARD.QUALITY_BONUSES.CREATOR_INFO;
    appliedBonuses.push('👤 Creator Info');
  }
  
  if (data.hashtags && data.hashtags.length >= 3) {
    breakdown.qualityBonuses.richHashtags = EARNINGS_STANDARD.QUALITY_BONUSES.RICH_HASHTAGS;
    appliedBonuses.push('#️⃣ Hashtags');
  }
  
  if (data.post_caption && data.post_caption.length > 10) {
    breakdown.qualityBonuses.caption = EARNINGS_STANDARD.QUALITY_BONUSES.CAPTION_PROVIDED;
    appliedBonuses.push('💬 Caption');
  }
  
  // Apply performance bonuses
  if (data.views_count) {
    if (data.views_count >= 1000000) {
      breakdown.performanceBonuses.viral = EARNINGS_STANDARD.PERFORMANCE_BONUSES.VIRAL_CONTENT;
      appliedBonuses.push('🔥 Viral (1M+ views)');
    } else if (data.views_count >= 100000) {
      breakdown.performanceBonuses.highViews = EARNINGS_STANDARD.PERFORMANCE_BONUSES.HIGH_VIEWS;
      appliedBonuses.push('👀 High Views (100k+)');
    }
  }
  
  if (data.views_count && data.likes_count) {
    const engagementRate = data.likes_count / data.views_count;
    if (engagementRate > 0.1) {
      breakdown.performanceBonuses.highEngagement = EARNINGS_STANDARD.PERFORMANCE_BONUSES.HIGH_ENGAGEMENT;
      appliedBonuses.push('💯 High Engagement');
    }
  }
  
  if (data.wave_score && data.wave_score > 70) {
    breakdown.performanceBonuses.highWave = EARNINGS_STANDARD.PERFORMANCE_BONUSES.HIGH_WAVE_SCORE;
    appliedBonuses.push('🌊 High Wave Score');
  }
  
  if (data.isFinanceTrend || 
      (data.tickers && data.tickers.length > 0) || 
      (data.category && ['finance', 'crypto', 'stocks', 'trading'].includes(data.category))) {
    breakdown.performanceBonuses.finance = EARNINGS_STANDARD.PERFORMANCE_BONUSES.FINANCE_TREND;
    appliedBonuses.push('📈 Finance Trend');
  }
  
  // Calculate totals
  const qualityBonusTotal = Object.values(breakdown.qualityBonuses).reduce((sum, val) => sum + val, 0);
  const performanceBonusTotal = Object.values(breakdown.performanceBonuses).reduce((sum, val) => sum + val, 0);
  
  const baseAmount = breakdown.base;
  const bonusAmount = qualityBonusTotal + performanceBonusTotal;
  const subtotal = baseAmount + bonusAmount;
  
  // Apply multipliers
  const tierMultiplier = breakdown.multipliers.tier;
  const streakMultiplier = breakdown.multipliers.streak;
  
  // Add multiplier descriptions
  if (spotterTier === 'elite') {
    appliedBonuses.push('🏆 Elite Tier (1.5x)');
  } else if (spotterTier === 'verified') {
    appliedBonuses.push('✅ Verified Tier (1.0x)');
  } else if (spotterTier === 'learning') {
    appliedBonuses.push('📚 Learning Tier (0.7x)');
  } else if (spotterTier === 'restricted') {
    appliedBonuses.push('⚠️ Restricted Tier (0.3x)');
  }
  
  if (streakMultiplier > 1) {
    appliedBonuses.push(`🔥 ${streakMultiplier}x Streak Bonus`);
  }
  
  let finalAmount = subtotal * tierMultiplier * streakMultiplier;
  
  // Apply cap
  if (finalAmount > EARNINGS_STANDARD.LIMITS.MAX_SINGLE_SUBMISSION) {
    finalAmount = EARNINGS_STANDARD.LIMITS.MAX_SINGLE_SUBMISSION;
    breakdown.capped = true;
    appliedBonuses.push('🔒 Capped at $3.00');
  }
  
  breakdown.total = finalAmount;
  
  return {
    baseAmount,
    bonusAmount,
    tierMultiplier,
    streakMultiplier,
    finalAmount,
    appliedBonuses,
    breakdown,
  };
}

/**
 * Calculate earnings for a validation vote
 */
export function calculateValidationEarnings(
  isCorrectVote: boolean = true,
  validatorTier: SpotterTier = 'learning'
): number {
  if (!isCorrectVote) return 0;
  
  const baseAmount = EARNINGS_STANDARD.BASE_RATES.VALIDATION_VOTE;
  const tierMultiplier = EARNINGS_STANDARD.TIER_MULTIPLIERS[validatorTier];
  
  return baseAmount * tierMultiplier;
}

/**
 * Calculate approval bonus for trend spotter
 */
export function calculateApprovalBonus(spotterTier: SpotterTier = 'learning'): number {
  const baseBonus = EARNINGS_STANDARD.BASE_RATES.APPROVAL_BONUS;
  const tierMultiplier = EARNINGS_STANDARD.TIER_MULTIPLIERS[spotterTier];
  
  return baseBonus * tierMultiplier;
}

/**
 * Get streak multiplier for given streak count
 */
export function getStreakMultiplier(streakCount: number): number {
  const multiplierKeys = Object.keys(EARNINGS_STANDARD.STREAK_MULTIPLIERS)
    .map(k => parseInt(k))
    .sort((a, b) => b - a);
  
  for (const key of multiplierKeys) {
    if (streakCount >= key) {
      return EARNINGS_STANDARD.STREAK_MULTIPLIERS[key as keyof typeof EARNINGS_STANDARD.STREAK_MULTIPLIERS];
    }
  }
  
  return 1.0;
}

/**
 * Check if user can cash out
 */
export function canCashOut(approvedBalance: number): boolean {
  return approvedBalance >= EARNINGS_STANDARD.LIMITS.MIN_CASHOUT_AMOUNT;
}

/**
 * Format currency for display
 */
export function formatEarnings(amount: number): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(amount);
}

/**
 * Get earning status label and styling
 */
export function getEarningStatusDisplay(status: EarningStatus) {
  const statusConfig = {
    pending: {
      label: 'Pending Verification',
      color: 'yellow',
      bgColor: 'bg-yellow-100',
      textColor: 'text-yellow-800',
      icon: '🟡',
      description: 'Awaiting community verification',
    },
    approved: {
      label: 'Approved',
      color: 'green',
      bgColor: 'bg-green-100',
      textColor: 'text-green-800',
      icon: '🟢',
      description: 'Verified and ready to cash out',
    },
    paid: {
      label: 'Paid Out',
      color: 'blue',
      bgColor: 'bg-blue-100',
      textColor: 'text-blue-800',
      icon: '🔵',
      description: 'Successfully cashed out',
    },
    cancelled: {
      label: 'Cancelled',
      color: 'gray',
      bgColor: 'bg-gray-100',
      textColor: 'text-gray-800',
      icon: '⚫',
      description: 'Trend rejected or cancelled',
    },
  };
  
  return statusConfig[status] || statusConfig.pending;
}

/**
 * Validate earning calculation matches expected range
 */
export function validateEarningAmount(amount: number, type: EarningType): boolean {
  switch (type) {
    case 'trend_submission':
      return amount >= 0 && amount <= EARNINGS_STANDARD.LIMITS.MAX_SINGLE_SUBMISSION;
    case 'trend_validation':
      return amount >= 0 && amount <= EARNINGS_STANDARD.BASE_RATES.VALIDATION_VOTE * 1.5;
    case 'approval_bonus':
      return amount >= 0 && amount <= EARNINGS_STANDARD.BASE_RATES.APPROVAL_BONUS * 1.5;
    default:
      return amount >= 0;
  }
}

// Export for backward compatibility
export const EARNINGS_CONFIG = EARNINGS_STANDARD;

File: backend/app/config/earnings_standard.py
"""
EARNINGS STANDARD - Python Implementation
Version: 1.0.0
Date: 2025-01-11

This file mirrors the TypeScript EARNINGS_STANDARD.ts configuration
to ensure consistency across frontend and backend systems.
"""

from decimal import Decimal
from enum import Enum
from typing import Dict, List, Optional, TypedDict
from dataclasses import dataclass


# ============================================
# ENUMS AND TYPE DEFINITIONS
# ============================================

class SpotterTier(str, Enum):
    MASTER = "master"
    ELITE = "elite"
    VERIFIED = "verified"
    LEARNING = "learning"
    RESTRICTED = "restricted"


class EarningType(str, Enum):
    TREND_SUBMISSION = "trend_submission"
    TREND_VALIDATION = "trend_validation"
    APPROVAL_BONUS = "approval_bonus"
    SCROLL_SESSION = "scroll_session"
    CHALLENGE = "challenge"
    REFERRAL = "referral"


class EarningStatus(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    PAID = "paid"
    CANCELLED = "cancelled"


@dataclass
class EarningBreakdown:
    base: Decimal
    quality_bonuses: Dict[str, Decimal]
    performance_bonuses: Dict[str, Decimal]
    tier_multiplier: Decimal
    streak_multiplier: Decimal
    total: Decimal
    capped: bool


@dataclass
class EarningCalculation:
    base_amount: Decimal
    bonus_amount: Decimal
    tier_multiplier: Decimal
    streak_multiplier: Decimal
    final_amount: Decimal
    applied_bonuses: List[str]
    breakdown: EarningBreakdown


# ============================================
# EARNINGS CONFIGURATION
# ============================================

class EarningsStandard:
    """Centralized earnings configuration matching EARNINGS_STANDARD.ts"""
    
    VERSION = "1.0.0"
    
    # Base earning rates (in USD)
    BASE_RATES = {
        "TREND_SUBMISSION": Decimal("0.25"),  # FIXED: Correct base rate
        "VALIDATION_VOTE": Decimal("0.10"),
        "APPROVAL_BONUS": Decimal("0.50"),
        "SCROLL_SESSION": Decimal("0.00"),
    }
    
    # Quality bonuses - REMOVED for simplicity
    # Earnings are now just: base × tier_multiplier × streak_multiplier
    QUALITY_BONUSES = {}
    
    # Performance bonuses - REMOVED for simplicity  
    # Earnings are now just: base × tier_multiplier × streak_multiplier
    PERFORMANCE_BONUSES = {}
    
    # Tier multipliers (matching database)
    TIER_MULTIPLIERS = {
        SpotterTier.MASTER: Decimal("3.0"),     # Master: 3x multiplier
        SpotterTier.ELITE: Decimal("2.0"),      # Elite: 2x multiplier
        SpotterTier.VERIFIED: Decimal("1.5"),   # Verified: 1.5x multiplier  
        SpotterTier.LEARNING: Decimal("1.0"),   # Learning: 1x multiplier (base)
        SpotterTier.RESTRICTED: Decimal("0.5"), # Restricted: 0.5x multiplier
    }
    
    # Streak multipliers (days of consecutive submissions)
    STREAK_MULTIPLIERS = {
        0: Decimal("1.0"),   # 0-1 days: 1x
        2: Decimal("1.2"),   # 2-6 days: 1.2x
        7: Decimal("1.5"),   # 7-13 days: 1.5x
        14: Decimal("2.0"),  # 14-29 days: 2x
        30: Decimal("2.5"),  # 30+ days: 2.5x
    }
    
    # System limits
    LIMITS = {
        "MAX_SINGLE_SUBMISSION": Decimal("3.00"),
        "MAX_DAILY_EARNINGS": Decimal("50.00"),
        "MIN_CASHOUT_AMOUNT": Decimal("5.00"),
        "STREAK_WINDOW_MINUTES": 5,
    }
    
    # Validation requirements
    VALIDATION = {
        "VOTES_TO_APPROVE": 2,
        "VOTES_TO_REJECT": 2,
        "MAX_VOTE_EARNINGS_DAILY": Decimal("10.00"),
    }
    
    # Processing settings
    PROCESSING = {
        "CASHOUT_HOURS": "24-48",
        "PAYMENT_METHODS": ["venmo", "paypal", "bank_transfer"],
    }


# ============================================
# CALCULATION FUNCTIONS
# ============================================

def get_streak_multiplier(streak_count: int) -> Decimal:
    """Get the multiplier for a given streak count"""
    multipliers = EarningsStandard.STREAK_MULTIPLIERS
    
    # Find the highest applicable multiplier
    for streak_threshold in sorted(multipliers.keys(), reverse=True):
        if streak_count >= streak_threshold:
            return multipliers[streak_threshold]
    
    return Decimal("1.0")


def calculate_trend_submission_earnings(
    trend_data: dict = None,
    spotter_tier: SpotterTier = SpotterTier.LEARNING,
    streak_count: int = 0
) -> EarningCalculation:
    """
    Calculate earnings for a trend submission
    Simple formula: $0.25 × tier_multiplier × streak_multiplier
    
    Args:
        trend_data: Dictionary containing trend submission data (not used in simple calculation)
        spotter_tier: The tier of the spotter
        streak_count: Current submission streak
    
    Returns:
        EarningCalculation with complete breakdown
    """
    
    # Get base amount
    base_amount = EarningsStandard.BASE_RATES["TREND_SUBMISSION"]
    
    # Get multipliers
    tier_multiplier = EarningsStandard.TIER_MULTIPLIERS[spotter_tier]
    streak_multiplier = get_streak_multiplier(streak_count)
    
    # Calculate final amount
    final_amount = base_amount * tier_multiplier * streak_multiplier
    
    # Build description
    applied_bonuses = []
    
    # Add tier description
    tier_descriptions = {
        SpotterTier.MASTER: f"👑 Master Tier ({tier_multiplier}x)",
        SpotterTier.ELITE: f"🏆 Elite Tier ({tier_multiplier}x)",
        SpotterTier.VERIFIED: f"✅ Verified Tier ({tier_multiplier}x)",
        SpotterTier.LEARNING: f"📚 Learning Tier ({tier_multiplier}x)",
        SpotterTier.RESTRICTED: f"⚠️ Restricted Tier ({tier_multiplier}x)",
    }
    applied_bonuses.append(tier_descriptions[spotter_tier])
    
    # Add streak bonus if applicable
    if streak_multiplier > Decimal("1"):
        applied_bonuses.append(f"🔥 {streak_count} day streak ({streak_multiplier}x)")
    
    # Initialize breakdown (simplified)
    breakdown = EarningBreakdown(
        base=base_amount,
        quality_bonuses={},  # No quality bonuses in simple model
        performance_bonuses={},  # No performance bonuses in simple model
        tier_multiplier=tier_multiplier,
        streak_multiplier=streak_multiplier,
        total=final_amount,
        capped=False
    )
    
    return EarningCalculation(
        base_amount=base_amount,
        bonus_amount=Decimal("0"),  # No bonuses in simple model
        tier_multiplier=tier_multiplier,
        streak_multiplier=streak_multiplier,
        final_amount=final_amount,
        applied_bonuses=applied_bonuses,
        breakdown=breakdown
    )


def calculate_validation_earnings(
    is_correct_vote: bool = True,
    validator_tier: SpotterTier = SpotterTier.LEARNING
) -> Decimal:
    """
    Calculate earnings for a validation vote
    
    Args:
        is_correct_vote: Whether the vote was correct
        validator_tier: The tier of the validator
    
    Returns:
        Earning amount
    """
    if not is_correct_vote:
        return Decimal("0")
    
    base_amount = EarningsStandard.BASE_RATES["VALIDATION_VOTE"]
    tier_multiplier = EarningsStandard.TIER_MULTIPLIERS[validator_tier]
    
    return base_amount * tier_multiplier


def calculate_approval_bonus(spotter_tier: SpotterTier = SpotterTier.LEARNING) -> Decimal:
    """
    Calculate approval bonus for trend spotter
    
    Args:
        spotter_tier: The tier of the spotter
    
    Returns:
        Bonus amount
    """
    base_bonus = EarningsStandard.BASE_RATES["APPROVAL_BONUS"]
    tier_multiplier = EarningsStandard.TIER_MULTIPLIERS[spotter_tier]
    
    return base_bonus * tier_multiplier


def can_cash_out(approved_balance: Decimal) -> bool:
    """Check if user can cash out their earnings"""
    return approved_balance >= EarningsStandard.LIMITS["MIN_CASHOUT_AMOUNT"]


def validate_earning_amount(amount: Decimal, earning_type: EarningType) -> bool:
    """Validate that an earning amount is within expected range"""
    
    if earning_type == EarningType.TREND_SUBMISSION:
        return Decimal("0") <= amount <= EarningsStandard.LIMITS["MAX_SINGLE_SUBMISSION"]
    elif earning_type == EarningType.TREND_VALIDATION:
        max_validation = EarningsStandard.BASE_RATES["VALIDATION_VOTE"] * Decimal("1.5")
        return Decimal("0") <= amount <= max_validation
    elif earning_type == EarningType.APPROVAL_BONUS:
        max_approval = EarningsStandard.BASE_RATES["APPROVAL_BONUS"] * Decimal("1.5")
        return Decimal("0") <= amount <= max_approval
    else:
        return amount >= Decimal("0")


def format_earnings(amount: Decimal) -> str:
    """Format earnings for display"""
    return f"${amount:.2f}"


# Export singleton instance
EARNINGS_STANDARD = EarningsStandard()